pwndbg/heap_tracking.py

```diff
@@ -0,0 +1,357 @@
+"""
+Heap Tracking
+
+This module implements runtime tracking of the heap, allowing pwndbg to detect
+heap related misbehavior coming from an inferior in real time, which lets us
+catch UAF bugs, double frees (and more), and report them to the user.
+
+# Approach
+The approach used starting with using breakpoints to hook into the following
+libc symbols: `malloc`, `free`, `calloc`, and `realloc`. Each hook has a
+reference to a shared instance of the `Tracker` class, which is responsible for
+handling the tracking of the chunks of memory from the heap.
+
+The tracker keeps two sorted maps of chunks, for freed and in use chunks, keyed
+by their base address. Newly allocated chunks are added to the map of in use
+chunks right before an allocating call returns, and newly freed chunks are moved
+from the map of in use chunks to the map of free ones right before a freeing
+call returns. The tracker is also responsible for installing watchpoints for
+free chunks when they're added to the free chunk map and deleting them when
+their corresponding chunks are removed from the map.
+
+Additionally, because going through the data structures inside of libc to
+determine whether a chunk is free or not is, more often than not, a fairly slow
+operation, this module will only do so when it determines its view of the chunks
+has diverged from the one in libc in a way that would affect behavior. When such
+a diffence is detected, this module will rebuild the chunk maps in the range it
+determines to have been affected.
+
+Currently, the way it does this is by deleting and querying from libc the new
+status of all chunks that overlap the region of a new allocation when it detects
+that allocation overlaps chunks it previously considered free.
+
+This approach lets us avoid a lot of the following linked lists that comes with
+trying to answer the allocation status of a chunk, by keeping at hand as much
+known-good information as possible about them. Keep in mind that, although it is
+much faster than going to libc every time we need to know the allocation status
+of a chunk, this approach does have drawbacks when it comes to memory usage.
+
+# Compatibility
+Currently module assumes the inferior is using GLibc.
+
+There are points along the code in this module where the assumptions it makes
+are explicitly documented and checked to be valid for the current inferior, so
+that it may be immediately clear to the user that something has gone wrong if
+they happen to not be valid. However, be aware that there may be assumptions
+that were not made explicit.
+
+"""
+
+import gdb
+from sortedcontainers import SortedDict
+
+import pwndbg.gdblib
+
+MALLOC_NAME = "malloc"
+FREE_NAME = "free"
+
+
+class FreeChunkWatchpoint(gdb.Breakpoint):
+    def __init__(self, chunk, tracker):
+        self.chunk = chunk
+        self.tracker = tracker
+        super().__init__(
+            f"*(char[{chunk.size}]*){chunk.address:#x}", type=gdb.BP_WATCHPOINT, internal=True
```

Comment:
nitpick: will it work if user did `set language rust` before running this?

Reply:
I've been testing out the extent of this patch's support for Rust with this small program:
```rust
fn main() {
    let a = Box::new(1u32);
    let a = Box::into_raw(a);

    let b = a as usize;
    unsafe {
        /* Fine. This object is still alive. */
        *a = 0;

        /* Free the object. */
        let _ = Box::from_raw(a);

        /* Trigger UAF.
         *
         * We do a round trip pointer conversion here to skid around provenance
         * rules that would otherwise apply if we were to read directly from the
         * original pointer, and trigger UB. We want UAF soup, not ASM pretzels.
         */
         let a = b as *mut u32;
         let _ = std::ptr::read_volatile(a);
    }
}
```

But it seems like the version of GDB I'm testing with has an issue in `gdb.FinishBreakpoint` when Rust frames get mixed with C ones (such as when the Rust stdlib calls `malloc`). In my testing, GDB calls `gdb.FinishBreakpoint.out_of_scope()` after malloc and free return, rather than `stop()`, as it does in an all-C codebase.

I also notice that GDB is reporting the wrong function names when it breaks in the main function.

Clearly supporting Rust goes deeper than just fixing fixing the expression.

Here are the versions of GDB, rustc and the command line used to build the test program:
```
pwndbg> version
Gdb:      9.2
Python:   3.8.10 (default, May 26 2023, 14:05:08)  [GCC 9.4.0]
Pwndbg:   2023.07.17 build: 49eac60
Capstone: 5.0.1280
Unicorn:  2.0.1
```

```
$ rustc --version
rustc 1.66.1 (90743e729 2023-01-10) (built from a source tarball)
```

```
$ rustc a/src/main.rs -C opt-level=3 -C lto=fat -C debuginfo=2 -o a.rust.x86
```